lib/AST/Bounds.cpp

```diff
@@ -474,9 +474,14 @@ class DeclBoundsFinder : public clang::DeclVisitor<DeclBoundsFinder>,
       }
 
       // NOTE(pag): This is a heuristic for detecting when things go "too far."
+      // NOTE(kumarak): We have gone too far and the token bound should end
+      //                if it reaches to a namespace. Return previous token in
+      //                that case
       if (tok_kind == clang::tok::kw_namespace) {
-        assert(kind == clang::tok::kw_namespace ||
-               tok[-1].Kind() == clang::tok::kw_using);
+        if (!(kind == clang::tok::kw_namespace ||
+            tok[-1].Kind() == clang::tok::kw_using)) {
```

Comment:
I think what's needed here is a helper method, like `PrevParsedToken`. It should return `nullptr` if there isn't a previous token. It should take in `tok`, and scan backwards, making sure to check if the token is actually a parsed token, i.e. `prev->IsParsed()`. Then, you can implement this check.

The key issue is that you could imagine a case involving some macro between `using` and `namespace`. If you notice other things that require looking at the previous token, then also update them to use this new method.

Having a `NextParsedToken` method would be good too.